### DISCUSSION TITLE
Compatibility of SofaCUDA with AttachConstraint and SurfacePressureConstraint from SoftRobots Plugin

### DISCUSSION THREAD

**Discussion by iman-adibnazari** [sofa-framework/sofa#4705] [SofaCUDA]
Hello,

I am trying to speedup simulations of a soft robot I'm working with (simplified version pictured below). The main approach I've been trying to take has been by integrating the SofaCUDA package where possible, but I seem to be running into some issues regarding constraints. I am trying to use some attach constraints between portions of the robot and I am getting errors that indicate that the CudaVec3 template might not be compatible with that component. I was wondering if there are going to be similar compatibility issues with the SurfacePressureConstraint from the SoftRobots plugin and if there are any recommended ways to navigate these issues. 

![image](https://github.com/sofa-framework/sofa/assets/18055974/50400473-330a-4f01-8372-4353a94190c8)
The above is an image of the robot while its deformed without any CUDA components. There are two low-stiffness segments (each with two internal pressure driven chambers) and one coupling component between the two that is higher-stiffness. 

![image](https://github.com/sofa-framework/sofa/assets/18055974/ad949848-3e31-48aa-ae62-0a97a019291e)
The above shows how the air chambers are configured in the neutral configuration of the robot, the purple points show points that I would like to be connected via AttachConstraint components, and the red box indicates a region of points that I'm currently tracking. 

I would greatly appreciate any insight or recommendations on how to speed up this simulation with SofaCUDA or any other tools as well.  Thank you!

**Comment by alxbilger** [comment]
Indeed, `AttachConstraint` has not been implemented in SofaCUDA, hence it cannot have a CUDA template. You can try to use a `IdentityMapping` to apply `AttachConstraint` on the CPU. This way your object is represented as a GPU-based MechanicalObject using the CUDA template, and a CPU-based MechanicalObject using a Vec3 template. The IdentityMapping links them both. Note that this solution implies CPU-GPU copies, hence it is not the most efficient implementation. But today, I am not sure you have other choices.

**Reply by iman-adibnazari** [reply]
That sounds great. Thank you for the idea! I will try it out.
